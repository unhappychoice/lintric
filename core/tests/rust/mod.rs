use insta::assert_snapshot;
use lintric_core::analyze_code;
use serde_json;

#[test]
fn test_analyze_code_basic() {
    let code = "
let a = 1;
let b = a + 1;
"
    .trim();
    let file_path = "test.rs".to_string();
    let result = analyze_code(code, file_path.clone(), file_path.clone()).unwrap();

    assert_snapshot!(serde_json::to_string_pretty(&result).unwrap());
}

#[test]
fn test_rust_function_call_dependency() {
    let code = "
fn add(a: i32, b: i32) -> i32 {
    a + b
}
fn main() {
    let x = add(1, 2);
}
"
    .trim();
    let file_path = "test.rs".to_string();
    let result = analyze_code(code, file_path.clone(), file_path.clone()).unwrap();

    assert_snapshot!(serde_json::to_string_pretty(&result).unwrap());
}

#[test]
fn test_rust_struct_field_access_dependency() {
    let code = r#"
struct Point { x: i32, y: i32 }
fn main() {
    let p = Point { x: 1, y: 2 };
    let val = p.x;
}
"#
    .trim();
    let file_path = "test.rs".to_string();
    let result = analyze_code(code, file_path.clone(), file_path.clone()).unwrap();

    assert_snapshot!(serde_json::to_string_pretty(&result).unwrap());
}

#[test]
fn test_rust_use_statements_dependency() {
    let code = r#"
mod my_module {
    pub struct MyStruct;
    pub fn my_function() {}
    pub const MY_CONST: i32 = 1;
}

use my_module::MyStruct;
use my_module::{my_function, MY_CONST};
use my_module::*;
use my_module as mm;

fn main() {
    let s = MyStruct;
    my_function();
    let c = MY_CONST;
    let s2 = mm::MyStruct;
}
"#
    .trim();
    let file_path = "test.rs".to_string();
    let result = analyze_code(code, file_path.clone(), file_path.clone()).unwrap();

    assert_snapshot!(serde_json::to_string_pretty(&result).unwrap());
}
