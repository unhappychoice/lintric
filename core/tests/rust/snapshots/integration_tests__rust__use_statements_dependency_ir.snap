---
source: core/tests/rust/mod.rs
expression: "serde_json::to_string_pretty(&ir).unwrap()"
---
{
  "file_path": "tests/rust/fixtures/use_statements_dependency.rs",
  "definitions": [
    {
      "name": "my_module",
      "line_number": 1,
      "definition_type": "ModuleDefinition",
      "scope": null
    },
    {
      "name": "MyStruct",
      "line_number": 2,
      "definition_type": "StructDefinition",
      "scope": "my_module.MyStruct"
    },
    {
      "name": "my_function",
      "line_number": 3,
      "definition_type": "FunctionDefinition",
      "scope": "my_module.my_function"
    },
    {
      "name": "MY_CONST",
      "line_number": 4,
      "definition_type": "ConstDefinition",
      "scope": "my_module"
    },
    {
      "name": "MyStruct",
      "line_number": 7,
      "definition_type": "ModuleDefinition",
      "scope": null
    },
    {
      "name": "main",
      "line_number": 12,
      "definition_type": "FunctionDefinition",
      "scope": "main"
    },
    {
      "name": "s",
      "line_number": 13,
      "definition_type": "VariableDefinition",
      "scope": "main"
    },
    {
      "name": "c",
      "line_number": 15,
      "definition_type": "VariableDefinition",
      "scope": "main"
    },
    {
      "name": "s2",
      "line_number": 16,
      "definition_type": "VariableDefinition",
      "scope": "main"
    }
  ],
  "dependencies": [
    {
      "source_line": 7,
      "target_line": 1,
      "symbol": "my_module",
      "dependency_type": "VariableUse",
      "context": "scoped_identifier"
    },
    {
      "source_line": 8,
      "target_line": 1,
      "symbol": "my_module",
      "dependency_type": "VariableUse",
      "context": "scoped_use_list"
    },
    {
      "source_line": 9,
      "target_line": 1,
      "symbol": "my_module",
      "dependency_type": "VariableUse",
      "context": "use_wildcard"
    },
    {
      "source_line": 10,
      "target_line": 1,
      "symbol": "my_module",
      "dependency_type": "VariableUse",
      "context": "use_as_clause"
    },
    {
      "source_line": 13,
      "target_line": 7,
      "symbol": "MyStruct",
      "dependency_type": "VariableUse",
      "context": "let_declaration"
    },
    {
      "source_line": 16,
      "target_line": 7,
      "symbol": "MyStruct",
      "dependency_type": "VariableUse",
      "context": "scoped_identifier"
    }
  ],
  "usage": [
    {
      "name": "my_module",
      "kind": "Identifier",
      "scope": null,
      "start_line": 7,
      "start_column": 5,
      "end_line": 7,
      "end_column": 14
    },
    {
      "name": "MyStruct",
      "kind": "Identifier",
      "scope": null,
      "start_line": 7,
      "start_column": 16,
      "end_line": 7,
      "end_column": 24
    },
    {
      "name": "my_module",
      "kind": "Identifier",
      "scope": null,
      "start_line": 8,
      "start_column": 5,
      "end_line": 8,
      "end_column": 14
    },
    {
      "name": "my_function",
      "kind": "Identifier",
      "scope": null,
      "start_line": 8,
      "start_column": 17,
      "end_line": 8,
      "end_column": 28
    },
    {
      "name": "MY_CONST",
      "kind": "Identifier",
      "scope": null,
      "start_line": 8,
      "start_column": 30,
      "end_line": 8,
      "end_column": 38
    },
    {
      "name": "my_module",
      "kind": "Identifier",
      "scope": null,
      "start_line": 9,
      "start_column": 5,
      "end_line": 9,
      "end_column": 14
    },
    {
      "name": "my_module",
      "kind": "Identifier",
      "scope": null,
      "start_line": 10,
      "start_column": 5,
      "end_line": 10,
      "end_column": 14
    },
    {
      "name": "mm",
      "kind": "Identifier",
      "scope": null,
      "start_line": 10,
      "start_column": 18,
      "end_line": 10,
      "end_column": 20
    },
    {
      "name": "MyStruct",
      "kind": "Identifier",
      "scope": "main",
      "start_line": 13,
      "start_column": 13,
      "end_line": 13,
      "end_column": 21
    },
    {
      "name": "my_function()",
      "kind": "CallExpression",
      "scope": "main",
      "start_line": 14,
      "start_column": 5,
      "end_line": 14,
      "end_column": 18
    },
    {
      "name": "my_function",
      "kind": "Identifier",
      "scope": "main",
      "start_line": 14,
      "start_column": 5,
      "end_line": 14,
      "end_column": 16
    },
    {
      "name": "MY_CONST",
      "kind": "Identifier",
      "scope": "main",
      "start_line": 15,
      "start_column": 13,
      "end_line": 15,
      "end_column": 21
    },
    {
      "name": "mm",
      "kind": "Identifier",
      "scope": "main",
      "start_line": 16,
      "start_column": 14,
      "end_line": 16,
      "end_column": 16
    },
    {
      "name": "MyStruct",
      "kind": "Identifier",
      "scope": "main",
      "start_line": 16,
      "start_column": 18,
      "end_line": 16,
      "end_column": 26
    }
  ],
  "analysis_metadata": {
    "language": "Rust",
    "total_lines": 17,
    "analysis_timestamp": "now",
    "lintric_version": "0.1.0"
  }
}
